pkg/pss/evaluate.go

```diff
@@ -0,0 +1,1104 @@
+	// to avoid panics when removing a new check.
+	removedChecks := 0
+	for checkIndex, check := range pssChecks {
+		for _, exclude := range rule.Exclude {
+			// Translate PSS control to check_id and remove it from PSSChecks if it's specified in exclude block
+			for _, CheckID := range PSS_controls_to_check_id[exclude.ControlName] {
+				if check.ID == CheckID && exclude.RestrictedField == "" && checkIndex <= len(pssChecks) {
+					index := checkIndex - removedChecks
+					pssChecks = append(pssChecks[:index], pssChecks[index+1:]...)
+					removedChecks++
+				}
+			}
+		}
+	}
+	return pssChecks
+}
+
+func forbiddenValuesExempted(ctx enginectx.Interface, pod *corev1.Pod, check PSSCheckResult, exclude kyvernov1.PodSecurityStandard, restrictedField string) (bool, error) {
+	if err := enginectx.AddJSONObject(ctx, pod); err != nil {
+		return false, errors.Wrap(err, "failed to add podSpec to engine context")
+	}
+
+	// spec.containers[*].securityContext.privileged
+	// -> spec.containers[?name=="nginx"].securityContext.privileged
+	value, err := ctx.Query(restrictedField)
+	if err != nil {
+		return false, errors.Wrap(err, fmt.Sprintf("failed to query value with the given path %s", exclude.RestrictedField))
+	}
+	if !allowedValues(value, exclude, PSS_controls[check.ID]) {
+		return false, nil
+	}
+	return true, nil
+}
+
+func checkContainerLevelFields(ctx enginectx.Interface, pod *corev1.Pod, check PSSCheckResult, exclude []kyvernov1.PodSecurityStandard, restrictedField restrictedField) (bool, error) {
+	if strings.Contains(restrictedField.path, "spec.containers[*]") {
+		for _, container := range pod.Spec.Containers {
+			matchedOnce := false
+			// Container.Name with double quotes
+			containerName := fmt.Sprintf(`"%s"`, container.Name)
+			if !strings.Contains(check.CheckResult.ForbiddenDetail, containerName) {
+				continue
+			}
+			for _, exclude := range exclude {
+				if !strings.Contains(exclude.RestrictedField, "spec.containers[*]") {
+					continue
+				}
+
+				//	Get values of this container only.
+				// spec.containers[*].securityContext.privileged -> spec.containers[?name=="nginx"].securityContext.privileged
+				newRestrictedField := strings.Replace(restrictedField.path, "*", fmt.Sprintf(`?name=='%s'`, container.Name), 1)
+
+				// No need to check if exclude.Images contains container.Image
+				// Since we only have containers matching the exclude.images with getPodWithMatchingContainers()
+
+				exempted, err := forbiddenValuesExempted(ctx, pod, check, exclude, newRestrictedField)
+				if err != nil || !exempted {
+					return false, nil
+				}
+				matchedOnce = true
+			}
+			// If container name is in check.Forbidden but isn't exempted by an exclude then pod creation is forbidden
+			if strings.Contains(check.CheckResult.ForbiddenDetail, container.Name) && !matchedOnce {
+				return false, nil
+			}
+		}
+	}
+	if strings.Contains(restrictedField.path, "spec.initContainers[*]") {
```

Comment:
The logic is more or less similar for all three types of containers. Can we extract the common logic with variants and reuse them?